src/txn/txn_ckpt.c

```diff
@@ -1523,9 +1524,26 @@ __checkpoint_mark_skip(WT_SESSION_IMPL *session, WT_CKPT *ckptbase, bool force)
     F_CLR(btree, WT_BTREE_SKIP_CKPT);
     if (!btree->modified && !force) {
         deleted = 0;
-        WT_CKPT_FOREACH (ckptbase, ckpt)
+        WT_CKPT_FOREACH (ckptbase, ckpt) {
+            /*
+             * Don't skip the objects that have obsolete pages to let them to be removed as part of
+             * checkpoint cleanup.
+             */
+            if (ckpt->size != 0) {
+                if (ckpt->ta.newest_stop_ts != WT_TS_MAX)
```

Comment:
A couple things:
1. It seems this code is specific enough that it is out of place in this function. Please make a small function that does the necessary checks and returns a boolean or something. Maybe `checkpoint_apply_obsolete` or something.
2. I am not understanding the new BTREE flag. It seems like the call to `wt_txn_visible_all` is the actual check for obsolete. What is the flag doing exactly? Add that as a comment in the new function.
3. Do we have any sense of how often this will trigger? Or, if it triggers once, and then the tree remains clean, will its next checkpoint have `ckpt->ta.newest_stop_ts` set again? We want to make sure that we don't undo the "skip parsing of clean trees" optimization.

Reply:
1. Added a function as suggested.
2. Skipping checkpoint is done in two places, one is at before retrieving the checkpoint list and another one is after retrieving the checkpoint list. The flag is used to avoid skipping the checkpoint if the btree has any obsolete pages. I added a comment.
3. For btree's that don't have completely removed don't have any problem with this code change. This change mainly affects the history store where it has a valid stop timestamp always. For the history store, we want to verify the checkpoint list before we skip it as part of the checkpoint. It doesn't trigger any checkpoints unnecessary until all the pages are obsolete.

I added more details to the jira ticket for future reference.